homeassistant/components/elkm1/__init__.py

```diff
@@ -63,40 +66,69 @@ def _elk_value(val):
+    schema = vol.Schema(vol.Unique())
+    schema(prefixes)
+    return value
+
+
+DEVICE_SCHEMA_SUBDOMAIN = vol.Schema({
     vol.Optional(CONF_ENABLED, default=True): cv.boolean,
     vol.Optional(CONF_INCLUDE, default=[]): [_elk_range_validator],
     vol.Optional(CONF_EXCLUDE, default=[]): [_elk_range_validator],
 })
 
+DEVICE_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PREFIX, default=''): cv.string,
+    vol.Optional(CONF_USERNAME, default=''): cv.string,
+    vol.Optional(CONF_PASSWORD, default=''): cv.string,
+    vol.Optional(CONF_TEMPERATURE_UNIT, default='F'):
+    cv.temperature_unit,
+    vol.Optional(CONF_AREA, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+    vol.Optional(CONF_COUNTER, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+    vol.Optional(CONF_KEYPAD, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+    vol.Optional(CONF_OUTPUT, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+    vol.Optional(CONF_PLC, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+    vol.Optional(CONF_SETTING, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+    vol.Optional(CONF_TASK, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+    vol.Optional(CONF_THERMOSTAT, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+    vol.Optional(CONF_ZONE, default={}): DEVICE_SCHEMA_SUBDOMAIN,
+}, _host_validator)
+
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: vol.Schema(
-        {
-            vol.Required(CONF_HOST): cv.string,
-            vol.Optional(CONF_USERNAME, default=''): cv.string,
-            vol.Optional(CONF_PASSWORD, default=''): cv.string,
-            vol.Optional(CONF_TEMPERATURE_UNIT, default='F'):
-                cv.temperature_unit,
-            vol.Optional(CONF_AREA, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-            vol.Optional(CONF_COUNTER, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-            vol.Optional(CONF_KEYPAD, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-            vol.Optional(CONF_OUTPUT, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-            vol.Optional(CONF_PLC, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-            vol.Optional(CONF_SETTING, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-            vol.Optional(CONF_TASK, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-            vol.Optional(CONF_THERMOSTAT, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-            vol.Optional(CONF_ZONE, default={}): CONFIG_SCHEMA_SUBDOMAIN,
-        },
-        _host_validator,
-    )
+    DOMAIN: vol.All(cv.ensure_list, _has_all_unique_prefixes,
+                    [DEVICE_SCHEMA])
 }, extra=vol.ALLOW_EXTRA)
 
 
+def _get_mac_for(elk_host):
+    hostname_pattern = re.compile(r"^elks?://([^:]+)")
+    digits_pattern = re.compile(r"^[0-9.]+$")
+    hp_m = hostname_pattern.search(elk_host)
+    if hp_m:
+        ip_host = hp_m.group(1)
+        dp_m = digits_pattern.match(ip_host)
+        if dp_m:
+            mac_addr = get_mac_address(ip=ip_host)
+        else:
+            mac_addr = get_mac_address(hostname=ip_host)
+        _LOGGER.info("elkm1 at %s (%s) has mac %s",
```

Comment:
Start logging messages with capital letter.

Reply:
I would change to debug level.

Reply:
I've been thinking more about the unique_id and I believe we should go back to using the prefix added in when it has been set rather than the mac address when we can get it.  Here are my arguments:

1) https://developers.home-assistant.io/docs/en/entity_index.html uses the language "should not be configurable" rather than must not, so an exception -- though rare -- must be permissible. (i.e., you can keep reading the rest of my arguments :-) ).

2) using prefix where legacy configurations lack the prefix configuration setting will mean that there is no breaking change.  This is potentially extra important for the elkm1 component because users are more likely to have renamed sensors from that component given the relatively short name length limit imposed by the elkm1.  Avoiding a breaking change is good.

3) the prefix can be documented as something that is meant to be a permanent unchanging description of that logical elkm1 controller.  E.g., if a controller dies and has to be swapped out, the prefix of the new one can be set the same to avoid losing the registry data. 

4) the prefix is guaranteed within executions to be unique (by the pre-config check that you proposed and I wrote).

5) Using the mac address read off of ARP on the network means that it's unstable during a network configuration change that happens at layer 2 or higher.  Realistically, a TCP proxy or ssh-forwarding would make the mac address unstable.

6) returning None results in unstable entity_ids for M1s which is completely unacceptable... one of the primary use cases for these sensors is automations for triggering lights, etc., and it'd be worse to have those break across reboots.

7) there is no better unique id accessible to the network interface of the device.

If you agree (or are willing to defer), I can prepare the change back to using the prefix and update the documentation to make the risks and arguments clear.

Thanks

Reply:
@balloob?

Reply:
I'd love your input, @balloob ... I'm excited to get this contribution finalized.

Reply:
Martin is right. Use mac. Prefix has to go. 

What is wrong with using element index for unique ID ?

Reply:
If you don't like mac, the alternative is to remove unique ID completely.